{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of VCF file containing variants along rAAV genome from datasets generated by SSV-Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create classes to handle Sample during VCF parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Sample:\n",
    "\n",
    "    def __init__ (self, id, min_var_freq, ref_seq):\n",
    "        self.id = id\n",
    "        self.min_var_freq = min_var_freq\n",
    "        self.position_list = []\n",
    "        \n",
    "        # iterate other bases of the reference sequence\n",
    "        for base in ref_seq:\n",
    "            \n",
    "            # Create a dict for all 4 DNA base where the reference base is identified        \n",
    "            self.position_list.append( {\n",
    "                \"A\": {\"ref_base\": True, \"freq\" : 1} if base == \"A\" else {\"ref_base\": False, \"freq\" : 0},\n",
    "                \"C\": {\"ref_base\": True, \"freq\" : 1} if base == \"C\" else {\"ref_base\": False, \"freq\" : 0},\n",
    "                \"G\": {\"ref_base\": True, \"freq\" : 1} if base == \"G\" else {\"ref_base\": False, \"freq\" : 0},\n",
    "                \"T\": {\"ref_base\": True, \"freq\" : 1} if base == \"T\" else {\"ref_base\": False, \"freq\" : 0}})\n",
    "            \n",
    "    def add_variant (self, pos, base, freq):\n",
    "        if freq > self.min_var_freq:   \n",
    "            self.position_list[pos-1][base][\"freq\"] = freq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test Sample and Position Classes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': 'test',\n",
       " 'min_var_freq': 0.01,\n",
       " 'position_list': [{'A': {'ref_base': True, 'freq': 1},\n",
       "   'C': {'ref_base': False, 'freq': 0},\n",
       "   'G': {'ref_base': False, 'freq': 0},\n",
       "   'T': {'ref_base': False, 'freq': 0}},\n",
       "  {'A': {'ref_base': False, 'freq': 0},\n",
       "   'C': {'ref_base': False, 'freq': 0},\n",
       "   'G': {'ref_base': False, 'freq': 0},\n",
       "   'T': {'ref_base': True, 'freq': 1}},\n",
       "  {'A': {'ref_base': True, 'freq': 0.8},\n",
       "   'C': {'ref_base': False, 'freq': 0.1},\n",
       "   'G': {'ref_base': False, 'freq': 0},\n",
       "   'T': {'ref_base': False, 'freq': 0.1}},\n",
       "  {'A': {'ref_base': False, 'freq': 0},\n",
       "   'C': {'ref_base': False, 'freq': 0},\n",
       "   'G': {'ref_base': False, 'freq': 0},\n",
       "   'T': {'ref_base': True, 'freq': 1}},\n",
       "  {'A': {'ref_base': False, 'freq': 0},\n",
       "   'C': {'ref_base': True, 'freq': 1},\n",
       "   'G': {'ref_base': False, 'freq': 0},\n",
       "   'T': {'ref_base': False, 'freq': 0}},\n",
       "  {'A': {'ref_base': False, 'freq': 0},\n",
       "   'C': {'ref_base': False, 'freq': 0},\n",
       "   'G': {'ref_base': True, 'freq': 1},\n",
       "   'T': {'ref_base': False, 'freq': 0}},\n",
       "  {'A': {'ref_base': True, 'freq': 1},\n",
       "   'C': {'ref_base': False, 'freq': 0},\n",
       "   'G': {'ref_base': False, 'freq': 0},\n",
       "   'T': {'ref_base': False, 'freq': 0}},\n",
       "  {'A': {'ref_base': False, 'freq': 0},\n",
       "   'C': {'ref_base': False, 'freq': 0},\n",
       "   'G': {'ref_base': False, 'freq': 0},\n",
       "   'T': {'ref_base': True, 'freq': 1}},\n",
       "  {'A': {'ref_base': False, 'freq': 0},\n",
       "   'C': {'ref_base': True, 'freq': 1},\n",
       "   'G': {'ref_base': False, 'freq': 0},\n",
       "   'T': {'ref_base': False, 'freq': 0}}]}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = Sample(\"test\", 0.01, \"ATATCGATC\")\n",
    "a.add_variant(3, 'T', 0.1)\n",
    "a.add_variant(3, 'C', 0.1)\n",
    "a.add_variant(3, 'A', 0.8)\n",
    "a.__dict__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parse the reference sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid mode: 'fasta'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-c74329c82c84>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mBio\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mref_seq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"/media/drew/easystore/umb_triley/ReferenceGenomes/UCSC_hg38/genome.fa\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mref_seq\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: invalid mode: 'fasta'"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "ref_seq = str(SeqIO.read(open(\"/media/drew/easystore/umb_triley/ReferenceGenomes/UCSC_hg38/genome.fa\",\"fasta\")).seq)\n",
    "ref_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define the min freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.001\n"
     ]
    }
   ],
   "source": [
    "min_freq = float(1)/1000\n",
    "print min_freq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define the Sample in VCF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_dict = {\n",
    " 'C1_AAV': Sample('C1_AAV', min_freq, ref_seq),\n",
    " 'RUN1_S7_AAV': Sample('RUN1_S7_AAV', min_freq, ref_seq),\n",
    " 'RUN2_S8_AAV': Sample('RUN2_S8_AAV', min_freq, ref_seq),\n",
    " 'RUN1_S2_AAV': Sample('RUN1_S2_AAV', min_freq, ref_seq),\n",
    " 'RUN2_S1_AAV': Sample('RUN2_S1_AAV', min_freq, ref_seq),\n",
    " 'RUN1_S4_AAV': Sample('RUN1_S4_AAV', min_freq, ref_seq),\n",
    " 'RUN2_S2_AAV': Sample('RUN2_S2_AAV', min_freq, ref_seq),\n",
    " 'RUN1_S6_AAV': Sample('RUN1_S6_AAV', min_freq, ref_seq),\n",
    " 'RUN2_S3_AAV': Sample('RUN2_S3_AAV', min_freq, ref_seq),\n",
    " 'RUN1_S8_AAV': Sample('RUN1_S8_AAV', min_freq, ref_seq),\n",
    " 'RUN2_S4_AAV': Sample('RUN2_S4_AAV', min_freq, ref_seq),\n",
    " 'RUN1_S3_AAV': Sample('RUN1_S3_AAV', min_freq, ref_seq),\n",
    " 'RUN2_S5_AAV': Sample('RUN2_S5_AAV', min_freq, ref_seq),\n",
    " 'RUN1_S5_AAV': Sample('RUN1_S5_AAV', min_freq, ref_seq),\n",
    " 'RUN2_S6_AAV': Sample('RUN2_S6_AAV', min_freq, ref_seq)}\n",
    " \n",
    "# 'RUN2_S7_AAV': \"Negative Control 2\", # Negative control have not interest for this analysis\n",
    "# 'RUN1_S1_AAV': \"Negative Control 1\", # Negative control have not interest for this analysis\n",
    "# 'RUN2_S9_AAV': Sample('', \"Internal Normalizer 3\", min_freq, ref_seq),"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Open vcf file and parse general metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: './NoIndel.vcf'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-52019a57fe1d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mHTSeq\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mvcfr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mHTSeq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mVCF_Reader\u001b[0m\u001b[0;34m(\u001b[0m \u001b[0;34m\"./NoIndel.vcf\"\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mvcfr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse_meta\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mvcfr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmake_info_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/HTSeq/__init__.py\u001b[0m in \u001b[0;36mparse_meta\u001b[0;34m(self, header_filename)\u001b[0m\n\u001b[1;32m    723\u001b[0m             \u001b[0mthe_iter\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mheader_filename\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"r\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    724\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 725\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0mline\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mthe_iter\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    726\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'#'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    727\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"##\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/HTSeq/utils.py\u001b[0m in \u001b[0;36m__iter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     36\u001b[0m                 \u001b[0mlines\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgzip\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfos\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'rt'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     37\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 38\u001b[0;31m                 \u001b[0mlines\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfos\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     39\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     40\u001b[0m             \u001b[0mlines\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfos\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: './NoIndel.vcf'"
     ]
    }
   ],
   "source": [
    "import HTSeq\n",
    "vcfr = HTSeq.VCF_Reader( \"./NoIndel.vcf\" ) \n",
    "vcfr.parse_meta() \n",
    "vcfr.make_info_dict() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parse VCF variants and add to Sample objects if an alternative base if found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "for variant in vcfr:\n",
    "    # Extract information for samples in sample dict\n",
    "    for name, sample in sample_dict.items():\n",
    "        \n",
    "        # Extract genotypes from the INFO field for the current sample\n",
    "        GT = [int(i) for i in variant.samples[name]['GT'].split('/')]\n",
    "        #print variant.samples[name]\n",
    "        # Extract more informations only if needed\n",
    "        if len(GT) > 1:\n",
    "            DPG = [int(i) for i in variant.samples[name]['DPG'].split(',')]\n",
    "            DP = int(variant.samples[name]['DP'])\n",
    "            base = [variant.ref]+variant.alt\n",
    "            \n",
    "            # Modifify sample for each alternative genotype\n",
    "            for i in range (len (GT)):\n",
    "                sample.add_variant(\n",
    "                    pos = variant.pos.pos,\n",
    "                    base = base[GT[i]],\n",
    "                    freq = float(DPG[i])/DP)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_group = {\n",
    "    \"Internal Normalizer\" : ['RUN1_S7_AAV', 'RUN2_S8_AAV'],}\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_group = {\n",
    "    \"IEX\" : ['RUN1_S3_AAV', 'RUN2_S5_AAV', 'RUN1_S5_AAV', 'RUN2_S6_AAV']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_group = {\n",
    "    \"Internal Normalizer\" : ['RUN1_S7_AAV', 'RUN2_S8_AAV'],\n",
    "    \"CsCl-\" : ['RUN1_S2_AAV', 'RUN2_S1_AAV'],\n",
    "    \"CsCl+\" : ['RUN1_S4_AAV', 'RUN2_S2_AAV'],\n",
    "    \"AVB-\" : ['RUN1_S6_AAV', 'RUN2_S3_AAV'],\n",
    "    \"AVB+\" : ['RUN1_S8_AAV', 'RUN2_S4_AAV'],\n",
    "    \"IEX-\" : ['RUN1_S3_AAV', 'RUN2_S5_AAV'],\n",
    "    \"IEX+\" : ['RUN1_S5_AAV', 'RUN2_S6_AAV'],\n",
    "    \"CsCl\" : ['RUN1_S2_AAV', 'RUN2_S1_AAV', 'RUN1_S4_AAV', 'RUN2_S2_AAV'],\n",
    "    \"AVB\" : ['RUN1_S6_AAV', 'RUN2_S3_AAV', 'RUN1_S8_AAV', 'RUN2_S4_AAV'],\n",
    "    \"IEX\" : ['RUN1_S3_AAV', 'RUN2_S5_AAV', 'RUN1_S5_AAV', 'RUN2_S6_AAV'],\n",
    "    \"ALL_AAV\" : ['RUN1_S2_AAV', 'RUN2_S1_AAV', 'RUN1_S4_AAV', 'RUN2_S2_AAV','RUN1_S6_AAV', 'RUN2_S3_AAV', 'RUN1_S8_AAV', 'RUN2_S4_AAV', 'RUN1_S3_AAV', 'RUN2_S5_AAV', 'RUN1_S5_AAV', 'RUN2_S6_AAV']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "for name, group in sample_group.items():\n",
    "    sample = Sample('name', min_freq, ref_seq)\n",
    "    \n",
    "    # For each position along the length of reference the sequence\n",
    "    for i in range (len(ref_seq)):\n",
    "        for base in [\"A\", \"C\", \"G\", \"T\"]:\n",
    "                        \n",
    "            # Not interested by the frequency of the reference base\n",
    "            if sample.position_list[i][base]['ref_base']:\n",
    "                sample.position_list[i][base][\"freq\"] = 0.0\n",
    "                \n",
    "            # Extract the list of frequency for this base in all samples\n",
    "            else: \n",
    "                var_freq_list = [sample_dict[var_name].position_list[i][base][\"freq\"] for var_name in group]\n",
    "                #print (var_freq_list)\n",
    "                \n",
    "                # At least half of the samples in the list must have the variant with a frequency > 0 else it is considered null\n",
    "                if len([j for j in var_freq_list if j != 0]) >= len(var_freq_list)/2:\n",
    "                    sample.position_list[i][base][\"freq\"] = sum(var_freq_list)/len(var_freq_list)\n",
    "                    #print (\"TRUE\")\n",
    "                else:\n",
    "                    sample.position_list[i][base][\"freq\"] = 0.0\n",
    "                \n",
    "    # Write 1 report per sample\n",
    "    with open (name+\"_report.csv\", \"wb\") as report:\n",
    "        report.write(\"Pos\\tAlt_A\\tAlt_C\\tAlt_G\\tAlt_T\\tsum\\n\")\n",
    "        for n, p in enumerate(sample.position_list):\n",
    "            report.write (\"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(\n",
    "                n, p[\"A\"]['freq'], p[\"C\"]['freq'], p[\"G\"]['freq'], p[\"T\"]['freq'],\n",
    "                p[\"A\"]['freq'] + p[\"C\"]['freq'] + p[\"G\"]['freq'] + p[\"T\"]['freq']))\n",
    "        \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
